benchmarks/ref/include/Bench.h

```diff
@@ -0,0 +1,91 @@
+//===- Bench.h - ----------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "Perf.h"
+#include "Tensor.h"
+#include <vector>
+
+// KernelInterface: Must be implemented by kernels to use Benchmark class.
+//
+// T is a numeric type (int, float, double), and represents the payload
+// type of the input and output tensors, not necessarily the accumulation
+// and working internal tensors.
+template <typename T> struct KernelInterface {
+  virtual ~KernelInterface() {}
+  // Runs the reference kernel with the list of arguments.
+  // Output needs to be passed in the args list and is up to the user to
+  // implement the correct logic.
+  virtual void runRef(std::vector<T> &args) = 0;
+};
+
+// Benchmark: runs a kernel multiple times, takes timings, return avg.
+//
+// Supports warm-up (JIT compilers, cache flush, etc), tensor initialization,
+// performance tracking, output comparison, etc.
+//
+// Usage:
+//  * Implement a kernel that implements the KernelInterface
+//  * Define a variable as Benchmark<MyKernel>
+template <class Kernel, class T> class Benchmark {
+  // Performance tracker, caches timings and calculates avg/dev
+  PerfResults perf;
+  // The kernel to execute, must implement KernelInterface
+  Kernel kernel;
+  // Cached arguments (and output) for the kernel call
+  std::vector<T> args;
+  // Number of times to run
+  size_t iter;
+  // Expected number of Giga FP ops in the kernel (for flops reporting)
+  double gflops;
+
+public:
+  Benchmark(size_t iter, double gflops = 0.0) : iter(iter), gflops(gflops) {}
+
+  // Sets argument (input/output) list.
+  void setArg(std::vector<T> &&arg) { args = std::move(arg); }
+
+  // Inspects the tensors
+  const T &getArg(size_t index) {
+    assert(index < args.size() && "Invalid index for arguments");
+    return args[index];
+  }
+
+  // Warms up the kernel and populates the output (inside the args vector).
+  void warmup() { kernel.runRef(args); }
+
+  // Runs the benchmark `iter` times and take timings
+  void run() {
+    // Easier to optimize / inline a branchless loop
+    for (size_t i = 0; i < iter; i++) {
+      perf.startTimer();
+      kernel.runRef(args);
```

Comment:
In a GPU, will this also time dispatch? If so, can you bundle the dispatch of multiple kernels in one go?

Reply:
It does include dispatch which I think is in line with MLIR measurements i.e., timing loop around kernel call that invokes `gpu.launch_func`.

Multiple kernels could be bundled or queued through streams but I haven't added this logic.